dev/src/document.ts

```diff
@@ -23,92 +23,56 @@ import {google} from '../protos/firestore_proto_api';
 
 /**
  * Returns a builder for DocumentSnapshot and QueryDocumentSnapshot instances.
  * Invoke `.build()' to assemble the final snapshot.
  *
  * @private
- * @class
  */
 export class DocumentSnapshotBuilder {
-  ref: DocumentReference;
-  fieldsProto: ApiMapValue;
-  readTime: Timestamp;
-  createTime: Timestamp;
-  updateTime: Timestamp;
+  /** The reference to the document. */
+  ref?: DocumentReference;
 
-  /**
-   * @private
-   * @hideconstructor
-   *
-   * @param {DocumentSnapshot=} snapshot An optional snapshot to base this
-   * builder on.
-   */
-  constructor(snapshot?) {
-    snapshot = snapshot || {};
+  /** The fields of the Firestore `Document` Protobuf backing this document. */
+  fieldsProto?: ApiMapValue;
 
-    /**
-     * The reference to the document.
-     *
-     * @type {DocumentReference}
-     */
-    this.ref = snapshot._ref;
+  /** The time when this document was read. */
+  readTime?: Timestamp;
 
-    /**
-     * The fields of the Firestore `Document` Protobuf backing this document.
-     *
-     * @type {object}
-     */
-    this.fieldsProto = snapshot._fieldsProto;
+  /** The time when this document was created. */
+  createTime?: Timestamp;
 
-    /**
-     * The time when this document was read.
-     *
-     * @type {Timestamp}
-     */
-    this.readTime = snapshot._readTime;
-
-    /**
-     * The time when this document was created.
-     *
-     * @type {Timestamp}
-     */
-    this.createTime = snapshot._createTime;
-
-    /**
-     * The time when this document was last updated.
-     *
-     * @type {Timestamp}
-     */
-    this.updateTime = snapshot._updateTime;
-  }
+  /** The time when this document was last updated. */
+  updateTime?: Timestamp;
 
   /**
    * Builds the DocumentSnapshot.
    *
    * @private
-   * @returns {QueryDocumentSnapshot|DocumentSnapshot} Returns either a
-   * QueryDocumentSnapshot (if `fieldsProto` was provided) or a
-   * DocumentSnapshot.
+   * @returns Returns either a QueryDocumentSnapshot (if `fieldsProto` was
+   * provided) or a DocumentSnapshot.
    */
   build(): QueryDocumentSnapshot|DocumentSnapshot {
     assert(
-        is.defined(this.fieldsProto) === is.defined(this.createTime),
+        (this.fieldsProto !== undefined) === (this.createTime !== undefined),
         'Create time should be set iff document exists.');
     assert(
-        is.defined(this.fieldsProto) === is.defined(this.updateTime),
+        (this.fieldsProto !== undefined) === (this.updateTime !== undefined),
         'Update time should be set iff document exists.');
     return this.fieldsProto ?
         new QueryDocumentSnapshot(
-            this.ref, this.fieldsProto, this.readTime, this.createTime,
-            this.updateTime) :
-        new DocumentSnapshot(this.ref, undefined, this.readTime);
+            this.ref!, this.fieldsProto!, this.readTime!, this.createTime!,
+            this.updateTime!) :
+        new DocumentSnapshot(this.ref!, undefined, this.readTime!);
```

Comment:
Just FYI- Using `!` in the process of making things more typed seems fine, but if this was new code, I'd ask if there's a way to rework this to avoid the need for `!` (either make the type non-undefined-able or add if-guards that guarantee it's defined).  I think ultimately we should be removing`!` and `any` usages over time.

Reply:
But it looks so much more important with all those `!!!`. I'll see if I can get rid of some of them soonish.